experimental/Pipeline/pipelie.mli

```diff
@@ -0,0 +1,204 @@
+
+(********************************************************************************)
+(*** Modules that belong to different files *)
+(********************************************************************************)
+module Feedback : sig
+  type t
+end
+
+module Trace : sig
+  type t
+end
+
+module Context : sig
+  type 'a t
+end
+
+module Sexp : sig
+  type t
+end
+
+module NameSymbol : sig
+  type t
+end
+
+(********************************************************************************)
+(** Meta Information *)
+(********************************************************************************)
+
+module Meta : sig
+  type t = { feedback : Feedback.t; trace : Trace.t }
+end
+
+
+(********************************************************************************)
+(** Step Information Information *)
+(********************************************************************************)
+
+module Step : sig
+  type 'a t =
+    | Success of {meta : Meta.t; result           : 'a}
+    | Failure of {meta : Meta.t; partial_feedback : 'a option}
+end
+
+(********************************************************************************)
+(** Pipeline *)
+(********************************************************************************)
+
+(** A Recursive list represents the concept of a list with a chance to nest
+ * This allows us to get a tree like structure,
+ * where A pipeline step can contain more steps
+ *)
+module RecursiveList : sig
+  type 'a t
+end
+
+(** A circular list represents the idea of steps that can be recursive
+ * on each other *)
+module CircularList : sig
+  (* TODO : add a recursion function to the recursive bit, that tells
+   * it how it's recursive. perhaps it uses *)
+  type 'a recursive_schema =
+    | Recursive of 'a list
+    | NonRecursive of 'a
+
+  type 'a t =
+    'a recursive_schema RecursiveList.t
+end
+
+module Pipeline : sig
+  (** [t] is the Pipeline argument itself. This is what is actively
+   * being processed in the pipeline itself.
+   *)
+  type t = {
+      current_expression : Sexp.t list;
+      context            : Sexp.t Context.t
+    }
+
+  type has_reduce = HasReduce
+
+  (* HasReduce m =>  *)
+  (** [step_function] represents the function type a pass must satisfy
+   * in the end. Note that the [Automation] module makes a more
+   * ergonomic interface to the [Pipeline]
+   *)
+  type step_function = t -> t Step.t
+
+  (** [named_step_function] is a named version of a function pass that
+   * is used to register the pass with the effect/system
+   *)
+  type named_step_function = NameSymbol.t * step_function
+
+  type eff = {
+      (* we keep the context as an effect as this is what will persist
+       * in the environment *)
+      context : Sexp.t Context.t;
+
+      current_expression : Sexp.t list;
+
+      (** [registered_pipeline] is the pipeline itself.
+       * It features all the steps named and registered with the system *)
+      registered_pipeline : named_step_function CircularList.t;
```

Comment:
Yes, if Haskell can satisfy the signature of a pass having an effect for this `eff` but I'm leaning on it can't express that with our current library choice. I'll have to think of it, but there is a plan `b` of having this env be taken by the env that is being run, so in theory it could, and it'll show up in a signature if this were to occur!